drivers/volume/portworx/portworx.go

```diff
@@ -388,6 +629,295 @@ func (p *portworx) VolumeDelete(pv *v1.PersistentVolume) error {
+			return false, nil
+		}
+
+		return true, nil
+	})
+	return snap, err
+}
+
+func (p *portworx) createVolumeSnapshotData(
+	snapshotName, snapshotNamespace string,
+	snapDataSource *crdv1.VolumeSnapshotDataSource,
+	snapStatus *[]crdv1.VolumeSnapshotCondition,
+	snapLabels map[string]string) (*crdv1.VolumeSnapshotData, error) {
+	var lastCondition crdv1.VolumeSnapshotDataCondition
+	if snapStatus != nil && len(*snapStatus) > 0 {
+		conditions := *snapStatus
+		ind := len(conditions) - 1
+		lastCondition = crdv1.VolumeSnapshotDataCondition{
+			Type:    (crdv1.VolumeSnapshotDataConditionType)(conditions[ind].Type),
+			Status:  conditions[ind].Status,
+			Message: conditions[ind].Message,
+		}
+	}
+
+	snapshotData := &crdv1.VolumeSnapshotData{
+		Metadata: metav1.ObjectMeta{
+			Name:   snapshotName,
+			Labels: snapLabels,
+		},
+		Spec: crdv1.VolumeSnapshotDataSpec{
+			VolumeSnapshotRef: &v1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Name:      snapshotName,
+				Namespace: snapshotNamespace,
+			},
+			VolumeSnapshotDataSource: *snapDataSource,
+		},
+		Status: crdv1.VolumeSnapshotDataStatus{
+			Conditions: []crdv1.VolumeSnapshotDataCondition{
+				lastCondition,
+			},
+		},
+	}
+
+	var result *crdv1.VolumeSnapshotData
+	err := wait.ExponentialBackoff(snapAPICallBackoff, func() (bool, error) {
+		var err error
+		result, err = k8s.Instance().CreateSnapshotData(snapshotData)
+		if err != nil {
+			logrus.Errorf("failed to create snapshotdata due to err: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if err != nil {
+		logrus.Errorf("Error creating the VolumeSnapshotData for snap %s: %v", snapshotName, err)
+		return nil, fmt.Errorf("Failed to create the VolumeSnapshotData for snap %s", snapshotName)
+	}
+
+	return result, nil
+}
+
+func (p *portworx) findParentPVCBySnapID(snapID string) (*v1.PersistentVolumeClaim, error) {
+	pxSnaps, err := p.volDriver.Inspect([]string{snapID})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(pxSnaps) != 1 {
+		return nil, fmt.Errorf("failed to inspect Portworx snapshot: %s. Received: %d entries in response",
+			snapID, len(pxSnaps))
+	}
+
+	pxSnap := pxSnaps[0]
+
+	if len(pxSnap.Source.Parent) == 0 {
+		return nil, fmt.Errorf("Portworx snapshot: %s does not have parent set", snapID)
+	}
+
+	snapParents, err := p.volDriver.Inspect([]string{pxSnap.Source.Parent})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(snapParents) != 1 {
+		return nil, fmt.Errorf("failed to inspect snapshot parent: %s. Received: %d entries in response",
+			pxSnap.Source.Parent, len(snapParents))
+	}
+
+	snapParent := snapParents[0]
+	parentPVName := snapParent.Locator.Name
+	parentPV, err := k8s.Instance().GetPersistentVolume(parentPVName)
+	if err != nil {
+		return nil, err
+	}
+
+	return k8s.Instance().GetPersistentVolumeClaim(parentPV.Spec.ClaimRef.Name, parentPV.Spec.ClaimRef.Namespace)
+}
+
+func (p *portworx) checkCloudSnapStatus(volID string) (string, error) {
```

Comment:
This method or the callers of the method also need to deal with "Failed" state. The callers should stop waiting in that case.